# E-01

(1) 필요한 모듈 임포트 하기

(2) 데이터 준비하기

(3) 데이터 이해하기
  - Feature, Label Data 지정하기
  - Target Names 출력하기
  - 데이터 Describe 하기
  
(4) train, test 데이터 분리하기

(5) 다양한 모델로 학습시키기
  - Decision Tree
  - Random Forest
  - SVM(Support Vector Machine)
  - SGD Classifier
  - Logistic Classifer
  
(6) 모델 평가하기
  - 평가지표 중 적절한 것은? 그리고 이유는?

# 평가지표

1. 3가지 데이터셋의 구성을 파악하고, 데이터를 이해하는 과정이 포함되어있는가?

(feature, label 선정을 위한 데이터 분석과정을 전개함)


2. 3가지 데이터셋에 대해 각각 5가지 모델을 이용하여 학습을 수행하고 결과값을 얻었는가?

(모델학습 및 테스트가 정상적으로 수행되고, 결과값을 얻었음)


3. 3가지 데이터셋에 대해 모델의 평가지표를 선택하고, 그 이유를 근거를 바탕으로 서술하였는가?

(모델학습을 통해 얻은 결과물을 바탕으로 평가지표를 선택하고, 본인의 의견을 서술하였음)

# 손글씨 분류하기

In [1]:
#필요한 라이브러리 불러오기
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [2]:
#데이터셋(digits data)를 불러오기
digits = load_digits()
print(digits.keys())

#해당 keys에서 필요한 data와 target(label)을 새로운 변수로 지정하기
digits_data = digits.data
digits_label = digits.target

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])


In [3]:
print(digits_data.shape)
print(digits_label.shape)

print(digits.target_names)

print(digits.DESCR)

(1797, 64)
(1797,)
[0 1 2 3 4 5 6 7 8 9]
.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the nu

데이터에 대해 한 번 살펴보자.
data의 개수는 총 1797개를 가지고 있으며, 우리가 원하는 target(label)은 0~9까지 존재한다.

In [4]:
#불러온 data와 target(label)을 train과 test로 분리하기(8:2 비율)  -> 가장 많이 이용하는 비율
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state = 15)

print('X_train 개수:', len(X_train), 'X_test 개수:', len(X_test))

X_train 개수: 1437 X_test 개수: 360


In [5]:
#Decision Tree를 이용하기
from sklearn.tree import DecisionTreeClassifier 

decision_tree = DecisionTreeClassifier(random_state=32) 
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.94      0.97        31
           1       0.82      0.82      0.82        38
           2       0.72      0.87      0.79        38
           3       0.85      0.81      0.83        27
           4       0.97      0.78      0.86        41
           5       0.82      0.89      0.85        35
           6       0.85      0.89      0.87        38
           7       0.91      0.91      0.91        34
           8       0.74      0.74      0.74        35
           9       0.83      0.79      0.81        43

    accuracy                           0.84       360
   macro avg       0.85      0.84      0.84       360
weighted avg       0.85      0.84      0.84       360



In [6]:
#Random Forest를 이용하기
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.94      0.95        31
           1       0.95      0.97      0.96        38
           2       1.00      1.00      1.00        38
           3       1.00      0.96      0.98        27
           4       0.95      1.00      0.98        41
           5       0.97      1.00      0.99        35
           6       1.00      0.95      0.97        38
           7       1.00      1.00      1.00        34
           8       0.94      0.97      0.96        35
           9       1.00      0.98      0.99        43

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [7]:
#SVM(Support Vector Machine)을 이용하기
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.98        31
           1       0.95      1.00      0.97        38
           2       1.00      1.00      1.00        38
           3       0.96      0.96      0.96        27
           4       0.98      0.98      0.98        41
           5       1.00      1.00      1.00        35
           6       1.00      1.00      1.00        38
           7       1.00      1.00      1.00        34
           8       0.97      0.94      0.96        35
           9       0.98      0.98      0.98        43

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [8]:
#SGD Classifier를 이용하기
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.97      0.97        31
           1       0.95      0.95      0.95        38
           2       0.97      0.97      0.97        38
           3       1.00      0.85      0.92        27
           4       0.98      0.98      0.98        41
           5       0.97      0.97      0.97        35
           6       0.95      1.00      0.97        38
           7       0.92      1.00      0.96        34
           8       0.82      0.94      0.88        35
           9       1.00      0.86      0.92        43

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



In [9]:
#Logistic Regression을 이용하기
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=5000)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.98        31
           1       0.95      0.97      0.96        38
           2       1.00      0.97      0.99        38
           3       0.96      0.93      0.94        27
           4       0.93      1.00      0.96        41
           5       0.94      0.97      0.96        35
           6       1.00      0.97      0.99        38
           7       1.00      1.00      1.00        34
           8       0.94      0.94      0.94        35
           9       0.98      0.95      0.96        43

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



Logistic Regression 함수의 기본 max_iter는 100인데 최대에 도달했다라는 warning이 발생하여 max_iter를 5000까지 증가시켜주었다.

결과를 한 번 살펴보자.
평가지표는 총 5가지(Accuracy, Precision, Recall, f1score, ros_auc_score)이 있다.
위 실험에서는 잘못 예측을 했다고해서 문제가 발생하지 않는다.

따라서 단순히 내가 얼마나 잘 맞췄는지가 중요하므로 평가지표로 Accuracy가 높아야 할 것으로 보인다.
그래서 ML 모델 중 RF와 SVM 모델이 가장 높은 Accuracy를 보여준다.


# 와인 분류하기

In [10]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [11]:
wines = load_wine()
print(wines.keys())

wines_data = wines.data
wines_label = wines.target

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])


In [12]:
print(wines_data.shape)
print(wines_label.shape)

print(wines.target_names)

print(wines.DESCR)

(178, 13)
(178,)
['class_0' 'class_1' 'class_2']
.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23 

In [13]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(wines_data, wines_label, test_size=0.2, random_state = 8)

print('X_train 개수:', len(X_train), 'X_test 개수:', len(X_test))
print('y_train 개수:', len(y_train), 'y_test 개수:', len(y_test))

X_train 개수: 142 X_test 개수: 36
y_train 개수: 142 y_test 개수: 36


In [14]:
from sklearn.tree import DecisionTreeClassifier 

decision_tree = DecisionTreeClassifier(random_state=32) 
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.91      0.91        11
           1       0.93      1.00      0.97        14
           2       1.00      0.91      0.95        11

    accuracy                           0.94        36
   macro avg       0.95      0.94      0.94        36
weighted avg       0.95      0.94      0.94        36



In [15]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00        11

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [16]:
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.82      0.82        11
           1       0.56      1.00      0.72        14
           2       0.00      0.00      0.00        11

    accuracy                           0.64        36
   macro avg       0.46      0.61      0.51        36
weighted avg       0.47      0.64      0.53        36



/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [17]:
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.91      0.80        11
           1       0.55      0.86      0.67        14
           2       0.00      0.00      0.00        11

    accuracy                           0.61        36
   macro avg       0.42      0.59      0.49        36
weighted avg       0.43      0.61      0.50        36



/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [18]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=5000)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.82      0.90        11
           1       0.82      1.00      0.90        14
           2       1.00      0.91      0.95        11

    accuracy                           0.92        36
   macro avg       0.94      0.91      0.92        36
weighted avg       0.93      0.92      0.92        36



5가지의 ML 기법 중 test 데이터의 개수가 적어 2 class가 하나도 없는 문제가 발생하였다.

그래서 train과 test dataset을 split 과정에서 random_state를 적절히 조정하면서 각 비율을 비슷하게 맞춰주었다.

해당 Dataset의 평가지표를 살펴보자.

 와인의 종류를 정확하게 예측하였는가인데 이 부분은 와인의 정보를 가지고 와인을 예측하는 부분인데 이것을 어느 문제를 해결하는데 이용을 하느냐가 중요한 포인트가 될 것 같다.
 
 하지만 일반적인 경우를 생각을 한다면 첫 문제인 숫자맞추기와 동일하다고 생각이 되어 단순히 얼마나 잘 맞추었는지 Accuracy가 중요한 지표가 될 것이다.


# 유방암 여부 진단

In [19]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [20]:
cancer = load_breast_cancer()
print(cancer.keys())

cancer_data = cancer.data
cancer_label = cancer.target

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [21]:
print(cancer_data.shape)
print(cancer_label.shape)

print(cancer.target_names)

print(cancer.DESCR)

(569, 30)
(569,)
['malignant' 'benign']
.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 featu

In [22]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(cancer_data, cancer_label, test_size=0.2, random_state = 20)

print('X_train 개수:', len(X_train), 'X_test 개수:', len(X_test))

X_train 개수: 455 X_test 개수: 114


In [23]:
from sklearn.tree import DecisionTreeClassifier 

decision_tree = DecisionTreeClassifier(random_state=32) 
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      0.92      0.94        48
           1       0.94      0.97      0.96        66

    accuracy                           0.95       114
   macro avg       0.95      0.94      0.95       114
weighted avg       0.95      0.95      0.95       114



In [24]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.94      0.97        48
           1       0.96      1.00      0.98        66

    accuracy                           0.97       114
   macro avg       0.98      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114



In [25]:
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.83      0.91        48
           1       0.89      1.00      0.94        66

    accuracy                           0.93       114
   macro avg       0.95      0.92      0.93       114
weighted avg       0.94      0.93      0.93       114



In [26]:
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.90      0.90      0.90        48
           1       0.92      0.92      0.92        66

    accuracy                           0.91       114
   macro avg       0.91      0.91      0.91       114
weighted avg       0.91      0.91      0.91       114



In [27]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=5000)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.94      0.94        48
           1       0.95      0.95      0.95        66

    accuracy                           0.95       114
   macro avg       0.95      0.95      0.95       114
weighted avg       0.95      0.95      0.95       114



마지막으로 유방암 데이터를 살펴보자.

이 문제는 사람의 건강과 직결된 문제로 잘못 예측을 했을 땐 문제가 된다. 여기서 암이 아닌 사람을 암이라 예측을 하는 것은 그래도 건강 상 문제는 없기 때문에 다행이라 생각할 수 있지만, 암이 걸린 사람을 잘못 예측을 하게 되면 문제가 발생한다.

 따라서 이 문제는 앞 선 2문제와는 달리 양성을 음성으로 잘못 예측하는 경우가 없어야 한다.평가지표는 Recall이어야 한다.
위의 모델 중 Recall 값이 높은 모델은 RF가 비교적 높은 값을 보이고 있다.